cmd/rocsp-tool/main.go

```diff
@@ -118,31 +154,285 @@ func main2() error {
+	if err != nil {
+		return nil, fmt.Errorf("couldn't setup database client: %s", err)
+	}
+
+	db.SetMaxOpenConns(dbConfig.MaxOpenConns)
+	db.SetMaxIdleConns(dbConfig.MaxIdleConns)
+	db.SetConnMaxLifetime(dbConfig.ConnMaxLifetime.Duration)
+	db.SetConnMaxIdleTime(dbConfig.ConnMaxIdleTime.Duration)
+	return db, nil
+}
+
+// setDefault sets the target to a default value, if it is zero.
+func setDefault(target *int, def int) {
+	if *target == 0 {
+		*target = def
+	}
+}
+
+type client struct {
+	issuers       []ShortIDIssuer
+	redis         *rocsp.WritingClient
+	db            *sql.DB // optional
+	ocspGenerator capb.OCSPGeneratorClient
+	clk           clock.Clock
+}
+
+func (cl *client) loadFromDB(ctx context.Context, speed ProcessingSpeed) error {
+	// To scan the DB efficiently, we want to select only currently-valid certificates. There's a
+	// handy expires index, but for selecting a large set of rows, using the primary key will be
+	// more efficient. So first we find a good id to start with, then scan from there. Note: since
+	// AUTO_INCREMENT can skip around a bit, we add padding to ensure we get all currently-valid
+	// certificates.
+	startTime := cl.clk.Now().Add(-24 * time.Hour)
+	var minID int64
+	err := cl.db.QueryRowContext(
+		ctx,
+		"SELECT MIN(id) FROM certificateStatus WHERE notAfter >= ?",
+		startTime,
+	).Scan(&minID)
+	if err != nil {
+		return fmt.Errorf("selecting minID: %w", err)
+	}
+
+	setDefault(&speed.RowsPerSecond, 2000)
+	setDefault(&speed.ParallelSigns, 100)
+
+	query := fmt.Sprintf("SELECT %s FROM certificateStatus WHERE id > ?",
+		strings.Join(sa.CertStatusMetadataFields(), ", "))
+	rows, err := cl.db.QueryContext(ctx, query, minID)
+	if err != nil {
+		return fmt.Errorf("scanning certificateStatus: %w", err)
+	}
+	defer rows.Close()
+
+	// Limit the rate of reading rows.
+	frequency := time.Duration(float64(time.Second) / float64(time.Duration(speed.RowsPerSecond)))
+	rowTicker := time.NewTicker(frequency)
+
+	statusesToSign := make(chan *sa.CertStatusMetadata)
+	successes := make(chan struct{}, speed.ParallelSigns)
+	errChan := make(chan error, speed.ParallelSigns)
+
+	var signersWg sync.WaitGroup
+
+	for i := 0; i < speed.ParallelSigns; i++ {
+		signersWg.Add(1)
+		go cl.signAndStoreResponses(statusesToSign, errChan, successes, &signersWg)
+	}
+	var n, errorCount int
+	// Consume all available successes and errors from the output channels.
+	// If none are immediately available, return.
+	consume := func() {
+		for {
+			select {
+			case <-successes:
+				n++
+			case err := <-errChan:
+				errorCount++
+				if errorCount < 10 {
+					log.Print(err)
+				} else if errorCount < 1000 && rand.Intn(1000) < 100 {
+					log.Print(err)
+				} else if errorCount < 100000 && rand.Intn(1000) < 10 {
+					log.Print(err)
+				} else if rand.Intn(1000) < 1 {
+					log.Print(err)
+				}
+			default:
+				return
+			}
+			if (n+errorCount)%10 == 0 {
+				log.Printf("stored %d response, %d errors", n, errorCount)
```

Comment:
Two nits:
1) rename `n` to `successCount`, because on first reading I thought it was a count of *all* attempts, not just successful ones, and that could lead to situations where this mod check is never true.
2) pluralize "responses"
```suggestion
			if (successCount+errorCount)%10 == 0 {
				log.Printf("stored %d responses, %d errors", successCount, errorCount)
```